### PR TITLE
Makecart

### PR DESCRIPTION
**Pull request by indigodarkwolf**
At long last, the `-cart` support and `makecart` utility for authoring proper .crt files. 

This can be broken down into three major pieces of work:

- Extracting and isolating cartridge logic so that cartridge.c/.h can be used by separate executables, without having to include all of x16emu.
- Adding all the cartridge header and file format tools into cartridge.c/.h
- Creating makecart to expose the cartridge header and file format tools via command-line options (and/or a config file of those command-line options).

`makecart` has an undocumented parameter `-testbins` which just spits out a bunch of `.bin` files containing a single repeating byte that identifies each `.bin`. If those are generated, then the included `test.cfg` file will generate a `testcart.crt` which can be loaded into x16emu with the `-cart` option, which itself now requires a `.crt` file beginning with a valid cartridge header.

The `cartridge_read()` function that reads cartridge data does not, at present, check for unpopulated banks in order to return `C0`, but that could be added. I would greatly prefer to do so in a separate PR, however, and with an otherwise-working cartridge image.

Here's the usage for `makecart`:
```
-help
        Print this message and exit.

-cfg <filename.cfg>
        Use this file to pack the cartridge data. Config file is simply the
        command line switches, one per line.

-desc "Name/Description"
        Set the description field of the cartridge file. This is limited to 32
        bytes and padded with null characters ('\0').

-author "Author Information"
        Set the author field of the cartridge file. This is limited to 32 bytes
        and padded with null characters ('\0').

-copyright "Copyright Information"
        Set the copyright field of the cartridge file. This is limited to 32
        bytes and padded with null characters ('\0').

-version "Version Information"
        Set the version field of the cartridge file. This is limited to 32 bytes
        and padded with null characters ('\0').

-fill <value>
        Set the fill value to use with any partially-filled banks of cartridge
        memory. Value can be defined in decimal, or in hexadecimal with a '$' or
        '0x' prefix. 8-bit values will be repeated every byte, 16-bit values every two
        bytes, and 32-bit values every 4 bytes.

-rom_file <start_bank> [<filenames.bin>...]
        Define rom banks from the specified list of files. File data is tightly
        packed -- if a file does not end on a 16KB interval, the next file will be inserted
        immediately after it within the same bank. If the last file does not end on a 16KB
        interval, the remainder of the rom will be filled with the value set by '-fill'.

-ram <start_bank> [<end bank>]
        Define one or more banks of uninitialized RAM.

-ram_file <start_bank> [<filenames.bin>...]
        Define pre-initialized ram banks from the specified list of files. File
        data is tightly packed like with -rom. If the last file does not end on a 16KB
        interval, the remainder of the rom will be filled with the value set by '-fill'.

-nvram <start_bank> [<end_bank>]
        Define one or more uninitialized nvram banks.

-nvram_file <start_bank> [<filenames.bin>...]
        Define pre-initialized nvram banks from the specified list of files. File
        data is tightly packed like with -rom. If the last file does not end on a 16KB
        interval, the remainder of the rom will be filled with the value set by '-fill'.

-nvram_value <start_bank> <end_bank>
        Define pre-initialized nvram banks with the value set by '-fill'.

-none <start_bank> [<end_bank>]
        Define one or more unpopulated banks of the cartridge. By default, all
        banks are unpopulated unless specified by a previous command-line option.

-o <output.crt>
        Set the filename of the output cartridge file.

-unpack <input.crt> [<rom_size>]
        Unpacks the binary data from the cartridge file into <rom_size> slices.
        (for use with an EPROM programmer.) The ouput files will be the same
        filename as the input file, with _### appended. This will also create a
        .cfg file that can be used to re-pack the files into a new CRT if needed.

All options can be specified multiple times, and are applied  in-order from
        left to right.
```

### CODE REVIEW

**Comment by indigodarkwolf** [comment]
Actually, the way memory.c is performing the check for whether to return `$C0` on open bus reads should be fine, and was part of the merge resolution.

File: src/cartridge.c [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+	} else {
+		memset(mem, 0, CART_BANK_SIZE);
+	}
+}
+
+bool 
+cartridge_load(const char *path, bool randomize)
+{
+	const char *extension = file_find_extension(path, NULL);
+	if(strncmp(extension, ".crt", 4) != 0) {
```
Comment (mooinglemur) [inline src/cartridge.c:33]: This case-sensitive comparison might be a cause of minor annoyance for people on case-preserving case-insensitive filesystems.  Just a small consideration.  I realize now this is probably also true for the compressed file extensions.

File: src/makecart.c [hunk 2]
```diff
@@ -0,0 +1,671 @@
+#define _POSIX_C_SOURCE 200809L
+
+#include "../src/cartridge.h"
+#include "../src/files.h"
```
Comment (mooinglemur) [inline src/makecart.c:4]: Was this path specification intentional?

File: src/memory.c [hunk 3]
```diff
@@ -379,18 +395,4 @@ emu_read(uint8_t reg, bool debugOn)
...
-
-	if (cartridge_file) {
-		CART = calloc(CART_SIZE, sizeof(uint8_t));
-		while (ret) {
-			ret = SDL_RWread(cartridge_file, &CART[i], 1, CART_SIZE - i);
-			i += ret;
-		}
-	}
-}
+}
```
Comment (mooinglemur) [inline src/memory.c:398]: I think with what was removed here, `memory.h` should be updated to match, perhaps adding the prototype for the added function here too.

**Review by mooinglemur — COMMENTED** [review]
The CI build.yml needs to be updated if we want to include the makecart binary, which I think we want to.

File: src/makecart.c (hunk 2 above)
Comment (indigodarkwolf) [reply src/makecart.c]: Whoopsie. This file had started life in another directory and was later moved back into ./src.

File: src/cartridge.c (hunk 1 above)
Comment (indigodarkwolf) [reply src/cartridge.c]: Worthwhile. Too bad strnicmp is a non-standard C function. Added a similar implementation that should provide a case-insensitive compare.

File: src/memory.c (hunk 3 above)
Comment (indigodarkwolf) [reply src/memory.c:398]: Aye, cleaned up memory.h and some symbols in main.c

**Review by mooinglemur — APPROVED** [review]

**Pull request merged**